### PR TITLE
Allocations

### PR DESCRIPTION
**Pull request by danielmarbach**
@Particular/nservicebus-maintainers as discussed in today's call here is the proposed change

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -29,6 +29,13 @@ top of things.
 * Run build.bat in the root to assure nothing else was accidentally broken.
 * We have a resharper layer that applies our coding standards so make sure that you're "all green in reshaper"
 
+## Guidelines
+
+Avoid allocations in hot paths:
+
+* Avoid LINQ.
+* Avoid using foreach over collections that do not have a struct enumerator.
```
Comment (timbussmann) [inline CONTRIBUTING.md]: do we need to be more precise here about the (internal) usage of collection interfaces (as they internally may provide a struct enumerator)?

**Comment by danielmarbach** [comment]
Good point. Can you add a suggestion?

**Comment by SeanFeldman** [comment]
I'm watching this thread and wondering - is this kind of guidance that is good for Core only, or should be a bit wider? A lot of good discussions and learning could be applied in the downstream projects, but they won't take place, and not as a discussion, since they are solely limited to this repo.

And could happen later, is what happened with Core/ASB. ASB maintainers lived the past with a container mindset, when container was long gone and discouraged. Unintentionally, we create silos.

**Comment by danielmarbach** [comment]
Originally we had the guidance in PlatformDev and we got encouraged to put
in into NSB repo...

File: CONTRIBUTING.md (hunk 1 above)
Comment (timbussmann) [inline CONTRIBUTING.md]: how about adding something like "avoid using collection interfaces internally"?

Comment (danielmarbach) [inline CONTRIBUTING.md]: @timbussmann  I updated the text slightly. What do you think?

**Comment by danielmarbach** [comment]
@timbussmann changed

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -29,6 +29,13 @@ top of things.
 * Run build.bat in the root to assure nothing else was accidentally broken.
 * We have a resharper layer that applies our coding standards so make sure that you're "all green in reshaper"
 
+## Guidelines
+
+Avoid allocations in hot paths:
+
+* Avoid LINQ.
+* Avoid using foreach over collections that do not have a struct enumerator. For public facing interfaces we favour readonly collections and enumerables. For internal types we favour speed and allocation reduction, therefore avoid using collection interfaces internally.
```
Comment (bording) [inline CONTRIBUTING.md]: Since we prefer american english, it should be "favor" instead of "favour"

**Comment by danielmarbach** [comment]
@bording feel free to change

**Pull request merged**